### ISSUE TITLE
CellOutputNodeBundle has output_count="0" in case of --build latexpdf

### ISSUE THREAD

**Issue by AakashGfude** [ExecutableBookProject/MyST-NB#151]
The issue is that,  the post-transform function - `CellOutputsToNodes`, to replace CellOutputBundleNode to docutil nodes according to mime-type, actually replaces CellOutputBundleNode with nothing(empty) in case of latexpdf build. Becuase of which, the pdf has no outputs from code-cells.

The reason being the `document` ast in this case has `<CellOutputBundleNode output_count="0"/>` in the `post-transform stage`, but in case of html is has correct ast with  `<CellOutputBundleNode output_count="1"/>` . 
While trying to do a traceback, I found that the `parser` itself seems to be working fine,  with ast being created there with  `<CellOutputBundleNode output_count="1"/>` as intended by this function 
```
# Render the Markdown tokens to docutils AST.
tokens_to_docutils(md_parser, env, tokens, document)
```

So, it seems like somewhere between parse and post-transform, there is something that is messing it up. 

I am trying to debug it, but would love insights on this @choldgraf @chrisjsewell @mmcky

**Comment by AakashGfude** [comment]
Ok, probably found out the reason for this. 
The `write` function in `LatexBuilder` of sphinx, has this function called `inline_all_toctrees` which basically merges individual doctrees into one. 

So, there is this `cast` function from `typing`  module inside this function 
```
tree = cast(nodes.document, tree.deepcopy())
```
which has this `deepcopy` function of python which basically converts 
`CellOutputBundleNode output_count="1"/></CellOutputNode>` to `<CellOutputBundleNode output_count="0"/></CellOutputNode>` . :(

Remember this function @mmcky , it is back again.

**Comment by choldgraf** [comment]
huh, that is a weird bug...maybe we run into this issue with the `pdfhtml` builder as well?

**Comment by mmcky** [comment]
Nice work @AakashGfude - back again. So the copy process just not be saving the metadata for the count.

**Comment by AakashGfude** [comment]
> huh, that is a weird bug...maybe we run into this issue with the `pdfhtml` builder as well?

Maybe not, because this function is used  in case of latex only.

> Nice work @AakashGfude - back again. So the copy process must not be saving the metadata for the count.

I guess so, maybe we can try a different format, to see if it the information persists. I will try out a few things.

**Comment by mmcky** [comment]
It will also affect singlehtml builder so will probably impact pdfhtml also.

**Comment by AakashGfude** [comment]
> It will also affect singlehtml builder so will probably impact pdfhtml also.

ooh yeah, the function is in `singlehtml` as well. missed that.

**Comment by AakashGfude** [comment]
So, the `deepcopy` function which is basically a function inherited from `docutils` as mentioned by @mmcky  here https://github.com/ExecutableBookProject/MyST-NB/issues/128 . reduces the count of `<CellOutputBundleNode output_count="1"/>` by 1. The `outputs` attached to `CellOutputBundleNode` is also stripped off, as `node.outputs` returns empty in the `CellOutputsToNodes` post-transform function. 
```
for node in self.document.traverse(CellOutputBundleNode):
   cell = {"outputs": node.outputs}
``` 

So I guess this stripping off of `outputs`  attribute of `CelllOutputBundleNode` is causing the decrease in count? 


Now, I tried testing this behaviour for `html` target as well, which does not go through that function during `jupyter-book` build, to verify if `deepcopy` is the culprit.

For one particular build, this line
```
for node in self.document.traverse(CellOutputBundleNode):
   cell = {"outputs": node.outputs}
``` 

was giving `[{'data': {'text/plain': '4.04891256782214e-16'}, 'execution_count': 2, 'metadata': {}, 'output_type': 'execute_result'}]
` as the result for `node.outputs`

When I changed that line to 
```
 for node in self.document.deepcopy().traverse(CellOutputBundleNode):
            cell = {"outputs": node.outputs}
```

**Notice the deepcopy.**

then it gave `' '` as the result for `node.outputs`.

So, deepcopy is indeed stripping off that attribute of this user-defined class
